### DISCUSSION TITLE
Controlling EB refinement

### DISCUSSION THREAD

**Discussion by serazo97** [AMReX-Combustion/PeleC#536] [Q&A] [answered]
Hi all, I'm working on some simulations and I'm having some trouble getting the refinement I want. Specifically, EBs seem to refine further from the boundaries and at finer levels than I need. I've been attempting to address this for a while but haven't had much success. For the refinement distance away from the EBs I thought the `amr.n_error_buf` option would help but no such luck when setting it to 1 or 0. For the refinement level, I haven't seen anything to help control this other than in the Tagging.H file where the option `max_eb_refine_lev` exists but doesn't seem to be implemented (or I'm using it wrong). I've attached an image showing what I mean. The left and right regions are solid and a detonation travels through the channel in the middle. The channel thickness is 4 blocks wide at the coarsest grid level. I don't know why the refining extends so far away from the edge of the EB both into the rest of the solid as well as into the open channel. It makes for that channel to have high refinement while nothing of interest is happening. The EB also lines up well with the coarser grid so this high level of refinement throughout the whole EB also seems unnecessary. Any help or info on this matter would be greatly appreciated! 
![image](https://user-images.githubusercontent.com/110516591/182525650-da9b4a15-965f-4bf8-a989-52170c1ce689.png)

**Comment by marchdf** [comment]
Hi, thanks for reaching out with your question! Typically to cap the level at which the EB is refined we would use something like this in the input file:
```
tagging.max_eb_refine_lev = 1
```
where the `1` indicates that we are only letting the EB be refined up to level 1. Sometimes, when the EBs are close to each other or to boundaries, we need to play with the following factor:
```
tagging.eb_detag_factor = 1.5
```
which controls how close/far the derefining of the EB happens. Can you try these options and let us know how it goes? It sounds like maybe you tried the first one already?

**Reply by serazo97** [reply]
Thanks for the quick reply! I had tried the first one previously but I tried them again just in case. The mesh still looks like the one I posted above. I don't know what good values are for the detagging factor, so I could just be off the mark and need to play with it some more. Sadly the refinement level capping definitely is not working for me right now. Here is my input file. In case it's relevant, I have tried setting `ebd.boundary_grad_stencil_type` to 1 and 0. I'm not sure what it does so I figured I'd try!

```# ------------------  INPUTS TO MAIN PROGRAM  -------------------
stop_time = 6
max_step = 300000
#amr.restart = /mmfs1/home/serazo/SFFP2/chk151000

# PROBLEM SIZE & GEOMETRY
geometry.is_periodic = 0 0 0
geometry.coord_sys   = 0  # 0 => cart, 1 => RZ  2=>spherical
geometry.prob_lo     =   0.0        0.0       0.0
geometry.prob_hi     =   26.0       1.0      60.0
amr.n_cell           =   52         2        120

#pelec.Riemann    = 0     # 0: HLL,  1: JBB,  2: HLLC
# >>>>>>>>>>>>>  BC KEYWORDS <<<<<<<<<<<<<<<<<<<<<<
# Interior, UserBC, Symmetry, SlipWall, NoSlipWall
# >>>>>>>>>>>>>  BC KEYWORDS <<<<<<<<<<<<<<<<<<<<<<
pelec.lo_bc       =  "NoSlipWall"  "NoSlipWall"  "NoSlipWall"
pelec.hi_bc       =  "NoSlipWall"  "NoSlipWall"  "FOExtrap"

# TIME STEP CONTROL
pelec.cfl            = 0.001     # cfl number for hyperbolic system
pelec.init_shrink    = 0.1     # scale back initial timestep
pelec.change_max     = 1.1     # scale back initial timestep
pelec.dt_cutoff      = 5.e-20  # level 0 timestep below which we halt

# DIAGNOSTICS & VERBOSITY
pelec.sum_interval = 1       # coarse time steps between computing mass on domain
pelec.v            = 1       # verbosity in PeleC cpp files
amr.v              = 1       # verbosity in Amr.cpp
#amr.grid_log       = grdlog  # name of grid logging file

# REFINEMENT / REGRIDDING 
amr.max_level       = 5       # maximum level number allowed
amr.ref_ratio       = 2 2 2 2 2 2 2 # refinement ratio
amr.regrid_int      = 2 2 2 2 2 2 2 # how often to regrid
amr.blocking_factor = 2 4 8       # block factor in grid generation
#amr.blocking_factor_x = 2 4 8      # block factor in grid generation
#amr.blocking_factor_y = 2 4 8
#amr.blocking_factor_z = 2 4 8
amr.max_grid_size   = 64
amr.n_error_buf     = 0        # number of buffer cells in error est
amr.grid_eff = 1

# CHECKPOINT FILES
amr.checkpoint_files_output = 1
amr.check_file              = /mmfs1/home/serazo/SFFP2/chk    # root name of checkpoint file
amr.check_int               = 1000    # number of timesteps between checkpoints

# PLOTFILES
amr.plot_files_output = 1
amr.plot_file         = /mmfs1/home/serazo/SFFP2/plt     # root name of plotfile
amr.plot_int          = 500   # number of timesteps between plotfiles


prob.pmf_datafile = "ffcmy-12-Drive.dat"


pelec.do_react = 1
pelec.chem_integrator = "ReactorRK64"
pelec.do_mol=1
pelec.diffuse_temp=1
pelec.diffuse_enth=1
pelec.diffuse_spec=1
pelec.diffuse_vel=1
pelec.sdc_iters = 2


tagging.dengrad = 1e-4
tagging.max_dengrad_lev = 2

pelec.flame_trac_name = CH2O
tagging.ftracerr = 1.0e-3
tagging.max_ftracerr_lev = 5

tagging.max_eb_refine_lev = 1
tagging.eb_detag_factor = 1.5

amr.derive_plot_vars = ALL
pelec.plot_massfrac = 1

eb2.geom_type = "test"
pelec.eb_boundary_T = 300.
pelec.eb_isothermal = 0
pelec.eb_noslip = 1
ebd.boundary_grad_stencil_type = 1

**Reply by serazo97** [reply]
I'm wondering now if the issue is the version of PeleC I'm using. It doesn't have the detagging criteria added, but it looks like the refinement level cap should work still. I'm going to clone the latest version and see if it works then.

**Reply by marchdf** [reply]
Yes, please use the latest. If that doesn't work, @esclapez might be able to help because he implemented this capability and probably has more insight.

**Reply by serazo97** [reply]
Well sadly it's still not working. Thanks for verifying that `tagging.max_eb_refine_lev` and `tagging.eb_detag_factor` are the correct flags to be using.

**Reply by esclapez** [reply]
I'm not entirely sure, but it looks like what you want is to have the EB refinement clustered more closely around the EB right ? This is mainly controlled by the 'properly nested' condition intrinsic to AMReX and is the result of the combination of a few parameters:
 - the `amr.blocking_factor`
 - the `amr.n_error_buf`
 - the `amr.grid_eff`
 To get tighter clustering of fine patches around EB, you should have small `amr.blocking_factor` and `amr.n_error_buf`, and a high `amr.grid_eff`.
As Marc mentioned, the detagging is used when the EB refinement level is smaller than the `amr.max_level` and you need to avoid having grid patches finer than you EB limit `tagging.max_eb_refine_lev` cross the EB. But that is not what you're trying to achieve at this point right ?

**Reply by serazo97** [reply]
@esclapez Thanks for the reply! I'm trying to have EB refinement clustered more closely around the EB as well as limit the refinement level for this cluster. What you said makes sense for the clustering and maybe there's not much more I can do beyond tweaking the domain to get smaller blocking factors. However, I feel like it's still possible to limit the level to which the EB itself refines some way that I'm missing. Using high `amr.grid_eff` most of the EB is not very refined, but at the boundaries it refines to the maximum AMR level when I feel that it could remain less refined. Maybe that also has to do with the 'properly nested' condition you mentioned, I'm not quite clear yet on how exactly the AMR works.

**Comment by serazo97** [comment, marked as answer]
Okay I think I was able to sort it out. As @esclapez pointed out, the clustering distance is intrinsic to AMReX and in my case I couldn't do much about it. However, I saw no reason why the refinement level cap should not be working correctly. After digging through some files, I found that the EB untagging distance is calculated by multiplying some terms, one of them being what looked to me like the `amr.n_error_buf` in the inputs file, which I had set to 0. Setting it to 1 and using the aforementioned `tagging.max_eb_refine_lev` and `tagging.eb_detag_factor`, I was able to finally limit the EB refinement level. Thanks to both of you for the very insightful comments!
